iceberg/iceberg-handler/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java

```diff
@@ -0,0 +1,336 @@
+      NestedField.optional(6, Column.POSITION_DELETE_RECORD_COUNT.name(), LongType.get()),
+      NestedField.optional(7, Column.POSITION_DELETE_FILE_COUNT.name(), IntegerType.get()),
+      NestedField.optional(8, Column.EQUALITY_DELETE_RECORD_COUNT.name(), LongType.get()),
+      NestedField.optional(9, Column.EQUALITY_DELETE_FILE_COUNT.name(), IntegerType.get()),
+      NestedField.optional(10, Column.TOTAL_RECORD_COUNT.name(), LongType.get()),
+      NestedField.optional(11, Column.LAST_UPDATED_AT.name(), LongType.get()),
+      NestedField.optional(12, Column.LAST_UPDATED_SNAPSHOT_ID.name(), LongType.get()));
+  }
+
+  /**
+   * Computes and writes the {@link PartitionStatisticsFile} for a given table's current snapshot.
+   *
+   * @param table The {@link Table} for which the partition statistics is computed.
+   * @return {@link PartitionStatisticsFile} for the current snapshot.
+   */
+  public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) {
+    return computeAndWriteStatsFile(table, null);
+  }
+
+  /**
+   * Computes and writes the {@link PartitionStatisticsFile} for a given table and branch.
+   *
+   * @param table The {@link Table} for which the partition statistics is computed.
+   * @param branch A branch information to select the required snapshot.
+   * @return {@link PartitionStatisticsFile} for the given branch.
+   */
+  public static PartitionStatisticsFile computeAndWriteStatsFile(Table table, String branch) {
+    Snapshot currentSnapshot = SnapshotUtil.latestSnapshot(table, branch);
+    if (currentSnapshot == null) {
+      Preconditions.checkArgument(
+          branch == null, "Couldn't find the snapshot for the branch %s", branch);
+      return null;
+    }
+
+    StructType partitionType = Partitioning.partitionType(table);
+    Schema schema = schema(partitionType);
+
+    Collection<PartitionStats> stats = PartitionStatsUtil.computeStats(table, currentSnapshot);
+    List<PartitionStats> sortedStats = PartitionStatsUtil.sortStats(stats, partitionType);
+    Iterator<PartitionStatsRecord> convertedRecords = statsToRecords(sortedStats, schema);
+    return writePartitionStatsFile(table, currentSnapshot.snapshotId(), schema, convertedRecords);
+  }
+
+  @VisibleForTesting
+  static PartitionStatisticsFile writePartitionStatsFile(
+      Table table, long snapshotId, Schema dataSchema, Iterator<PartitionStatsRecord> records) {
+    OutputFile outputFile = newPartitionStatsFile(table, snapshotId);
+    FileWriterFactory<Record> factory =
+        GenericFileWriterFactory.builderFor(table)
+        .dataSchema(dataSchema)
+        .dataFileFormat(fileFormat(outputFile.location()))
+        .build();
+    DataWriter<Record> writer =
+        factory.newDataWriter(
+        EncryptedFiles.encryptedOutput(outputFile, EncryptionKeyMetadata.EMPTY),
+        PartitionSpec.unpartitioned(),
+        null);
+    try (Closeable toClose = writer) {
```

Comment:
I think this approach can lead to errors. Think about writer is closed after this try, but later it can be still used in this method. I see now it is not used, but later it can be changed, refactored etc. and can lead to potential errors.

Reply:
same appoach is used everywhere in iceberg repo.